api/api_ethereum.go

```diff
@@ -1442,3 +1430,101 @@ func checkTxFee(gasPrice *big.Int, gas uint64, cap float64) error {
+	if vmerr != nil {
+		result.Error = vmerr.Error()
+	}
+	return result, nil
+}
+
+// CreateAccessList creates an EIP-2930 type AccessList for the given transaction.
+// Reexec and BlockNrOrHash can be specified to create the accessList on top of a certain state.
+func (api *EthereumAPI) CreateAccessList(ctx context.Context, args EthTransactionArgs, blockNrOrHash *rpc.BlockNumberOrHash) (interface{}, error) {
+	return doCreateAccessList(ctx, api.publicKlayAPI.b, args, blockNrOrHash)
+}
+
+// AccessList creates an access list for the given transaction.
+// If the accesslist creation fails an error is returned.
+// If the transaction itself fails, an vmErr is returned.
+func AccessList(ctx context.Context, b Backend, blockNrOrHash rpc.BlockNumberOrHash, args EthTransactionArgs) (acl types.AccessList, gasUsed uint64, vmErr error, err error) {
+	// Retrieve the execution context
+	db, header, err := b.StateAndHeaderByNumberOrHash(ctx, blockNrOrHash)
+	if db == nil || err != nil {
+		return nil, 0, nil, err
+	}
+	gasCap := uint64(0)
+	if rpcGasCap := b.RPCGasCap(); rpcGasCap != nil {
+		gasCap = rpcGasCap.Uint64()
+	}
+
+	// Ensure any missing fields are filled, extract the recipient and input data
+	if err := args.setDefaults(ctx, b); err != nil {
+		return nil, 0, nil, err
+	}
+	var to common.Address
+	if args.To != nil {
+		to = *args.To
+	} else {
+		to = crypto.CreateAddress(args.from(), uint64(*args.Nonce))
+	}
+	// Retrieve the precompiles since they don't need to be added to the access list
+	rules := b.ChainConfig().Rules(header.Number)
+	precompiles := vm.ActivePrecompiles(rules)
+
+	// Create an initial tracer
+	prevTracer := vm.NewAccessListTracer(nil, args.from(), to, precompiles)
+	if args.AccessList != nil {
+		prevTracer = vm.NewAccessListTracer(*args.AccessList, args.from(), to, precompiles)
+	}
+	for {
+		// Retrieve the current access list to expand
+		accessList := prevTracer.AccessList()
+		logger.Trace("Creating access list", "input", accessList)
+
+		// Copy the original db so we don't modify it
+		statedb := db.Copy()
+		// Set the accesslist to the last al
+		args.AccessList = &accessList
+		intrinsicGas, err := types.IntrinsicGas(args.data(), nil, args.To == nil, rules)
+		if err != nil {
+			return nil, 0, nil, err
+		}
+		msg, err := args.ToMessage(gasCap, header.BaseFee, intrinsicGas)
+		if err != nil {
+			return nil, 0, nil, err
+		}
+
+		// Apply the transaction with the access list tracer
+		tracer := vm.NewAccessListTracer(accessList, args.from(), to, precompiles)
+		config := vm.Config{Tracer: tracer, Debug: true}
+		vmenv, _, err := b.GetEVM(ctx, msg, statedb, header, config)
+		res, err := blockchain.ApplyMessage(vmenv, msg)
+		if err != nil {
+			tx, _ := args.toTransaction()
+			return nil, 0, nil, fmt.Errorf("failed to apply transaction: %v err: %v", tx.Hash(), err)
+		}
+		if tracer.Equal(prevTracer) {
+			return accessList, res.UsedGas, res.Unwrap(), nil
```

Comment:
I tested below, but with revert error, it does not return accessList. The original PR seems return accessList when it's revert error. Could you check one more time? rpc return code does not work as expected..? (maybe I might have tested wrong)
```
# test contract
pragma solidity >=0.8.0 <0.9.0;

contract CreateAccessListTestContract {
    uint256 storage1;
    address storage2;
    mapping (address => uint256) storage3;
    function successTest() public {
        storage3[storage2] = storage1;
    }
    function revertTest() public {
        if (storage3[storage2] == storage1) {
            revert("reverted for test");
        } else {
            storage3[storage2] = storage1;
        }
    }
    function get(address _addr) public view returns (uint256) {
        return storage3[_addr];
    }
}
```
```
# succeed transaction: return accessList
> klay.createAccessList({"from":klay.accounts[0], "to": "0xF803d0E974c8c87B4d6cbf296e30aD8E4F3cDC74", "data": "0xd975e40c"}, "latest")
{
  accessList: [{
      address: "0xf803d0e974c8c87b4d6cbf296e30ad8e4f3cdc74",
      storageKeys: ["0x0000000000000000000000000000000000000000000000000000000000000000", "0x0000000000000000000000000000000000000000000000000000000000000001", "0xac33ff75c19e70fe83507db0d683fd3465c996598dc972688b7ace676c89077b"]
  }],
  gasUsed: "0x55f6"
}

```
```
# reverted transaction: does not return accessList
> klay.createAccessList({"from":klay.accounts[0], "to": "0xF803d0E974c8c87B4d6cbf296e30aD8E4F3cDC74", "data": "0x81bd24d1"}, "latest")
Error: execution reverted: reverted for test
        at web3.js:6810:9(39)
        at send (web3.js:5221:62(29))
        at <eval>:1:22(13)
```

Reply:
Thank you! Now it should work for two cases, insufficient balance and reverted execution. Could you double-check it out?

Reply:
Those two cases works well now. Thank you